playbot/playbot.py

```diff
@@ -59,48 +60,55 @@ def name(self):
     def login(self):
         if not self.game:
             return
-        sleep(0.2)
+        sleep(0.1)
         self.send('takeover')
-        sleep(0.2)
+        sleep(0.1)
         data = self.get_data()
-        if b'Name?' in data:
+        if 'Name?' in data:
             self.name()
         else:
             self.game = False
 
     def heartbeat_socket(self):
-        if self.heartbeat > 1500:
+        if self.heartbeat > 15000:
             self.send('ack')
             self.heartbeat = 0
         else:
             self.heartbeat += 1
 
     def get_data(self):
+        buffor = ''
         try:
-            self.heartbeat_socket()
-            data = self.socket.recv(1024)
-            return data
-        except socket.timeout:
-            return b''
+            while True:
+                self.heartbeat_socket()
+                buffor += self.socket.recv(1).decode('utf-8')
+                if '\x04' in buffor:
+                    self.heartbeat = 0
+                    return buffor.split('\x04')[0]
+        except BlockingIOError:
+            self.game = True
```

Comment:
If there is IOException should not it be set to False ?
If I get error on read data then game should not continiue.

Reply:
`IOException` is called when connection is set as `non-blocking` and there are no data to receive from socket.
But it is highly possible that data will be inside socket buffer next time we ask.
Nothing to change here.